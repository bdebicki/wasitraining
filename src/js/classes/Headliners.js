import AbstractLineup from './AbstractLineup';
import { ARTIST_KEYS } from '../enums/artist';

/**
 * sort types:
 *	- false - don't sort artists
 *	- 'alphabeticalExceptHeadliners' - sort alphabetical artists from all groups except headliners
 *	- 'customOrder' - sort by 'order' props from artist object
 *	- 'customOrderExceptHeadliners' - sort by 'order' props from artist object but don't touch headliners
 */

export default class Headliners extends AbstractLineup {
	get headliners() {
		if (this.sortType === 'customOrder') {
			return this.sortOrderedHeadliners();
		}
		// sortType is false, 'customOrderExceptHeadliners' or 'alphabeticalExceptHeadliners'
		return this.notSortedHeadliners();
	}

	getFlatHeadlinersList() {
		let flatHeadliners = [];

		flatHeadliners = this.rawLineup
			.reduce((acc, day) => { // push headliners from days
				if (day.headliners) { // check does headliners was on that day
					acc.push(day.headliners);
				}

				return acc;
			}, [])
			.reduce((acc, artist) => acc.concat(artist), []); // flat array with headliners (remove nester arrays)

		return flatHeadliners;
	}

	notSortedHeadliners() { // merge headliners from all days into one flat array (with artists only)
		let headliners = [];

		headliners = this.getFlatHeadlinersList()
			.map((headlinersArtist) => { // change object artists to sting
				if (typeof headlinersArtist === 'object') {
					return headlinersArtist[ARTIST_KEYS.ARTIST];
				}

				return headlinersArtist;
			});

		return headliners;
	}

	sortOrderedHeadliners() {
		let headliners = [];

		headliners = this.getFlatHeadlinersList()
			.sort((a, b) => a.order - b.order) // sort headliners by order property
			.map((artist) => artist[ARTIST_KEYS.ARTIST]); // flattening array - remove objects and displays only artists

		return headliners;
	}
}
