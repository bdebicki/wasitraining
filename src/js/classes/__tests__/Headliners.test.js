import editionAlphabeticalExceptHeadliners from '../../../../tests/__mocks__/edition-alphabeticalExceptHeadliners-noMergeArtists.json';
import editionCustomOrder from '../../../../tests/__mocks__/edition-customOrder-mergeArtists.json';
import editionCustomOrderExceptHeadliners from '../../../../tests/__mocks__/edition-customOrderExceptHeadliners-mainByDaysAndMergeRest.json';
import editionNotSorted from '../../../../tests/__mocks__/edition-notSort-customLevels.json';
import Headliners from '../Headliners';

describe('tests headliners class', () => {
	it('return headliners sorted by alphabeticalExceptHeadliners type', () => {
		// having
		const headliners = new Headliners(editionAlphabeticalExceptHeadliners);

		// when
		const sortedHeadliners = headliners.headliners;

		// then
		const expectedHeadliners = ['Radiohead', 'Foo Fighters', 'The Weekend', 'The XX', 'Lorde'];
		expect(sortedHeadliners).toEqual(expectedHeadliners);
	});
	it('return headliners sorted by customOrder type', () => {
		// having
		const headliners = new Headliners(editionCustomOrder);

		// when
		const sortedHeadliners = headliners.headliners;

		// then
		const expectedHeadliners = ['BjÃ¶rk', 'Beastie Boys', 'Muse'];
		expect(sortedHeadliners).toEqual(expectedHeadliners);
	});
	it('return headliners sorted by customOrderExceptHeadliners type', () => {
		// having
		const headliners = new Headliners(editionCustomOrderExceptHeadliners);

		// when
		const sortedHeadliners = headliners.headliners;

		// then
		const expectedHeadliners = [
			'Snoop Dogg',
			'Fatboy Slim',
			'Faithless',
			'The White Stripes',
			'Lauryn Hill',
			'Underworld',
		];
		expect(sortedHeadliners).toEqual(expectedHeadliners);
	});
	it('return headliners not sorted', () => {
		// having
		const headliners = new Headliners(editionNotSorted);

		// when
		const sortedHeadliners = headliners.headliners;

		// then
		const expectedHeadliners = ['Pink', 'Cypress Hill', 'Massive Attack', 'Goldfrapp'];
		expect(sortedHeadliners).toEqual(expectedHeadliners);
	});
});
