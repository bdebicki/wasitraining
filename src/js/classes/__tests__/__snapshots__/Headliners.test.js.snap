// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests headliners class return headliners not sorted 1`] = `
Array [
  Object {
    "artist": "Pink",
    "day": 1,
    "displayName": "P!nk",
    "firstOnLine": Array [
      "headliners",
      "lineup",
    ],
    "line": 2,
  },
  Object {
    "artist": "Cypress Hill",
    "day": 1,
    "line": 2,
  },
  Object {
    "artist": "Massive Attack",
    "day": 2,
    "firstOnLine": Array [
      "headliners",
      "lineup",
    ],
    "line": 2,
  },
  Object {
    "artist": "Goldfrapp",
    "day": 2,
    "line": 2,
  },
]
`;

exports[`tests headliners class return headliners sorted by alphabeticalExceptHeadliners type 1`] = `
Array [
  Object {
    "artist": "The Black Keys",
    "day": 1,
  },
  Object {
    "artist": "Pearl Jam",
    "day": 2,
  },
  Object {
    "artist": "Jack White",
    "day": 3,
  },
  Object {
    "artist": "Faith No More",
    "day": 4,
  },
]
`;

exports[`tests headliners class return headliners sorted by customOrder type 1`] = `
Array [
  Object {
    "artist": "Bj√∂rk",
    "day": 2,
    "firstOnLine": Array [
      "headliners",
      "lineup",
    ],
    "order": 1,
  },
  Object {
    "artist": "Beastie Boys",
    "day": 1,
    "order": 2,
  },
  Object {
    "artist": "Muse",
    "day": 1,
    "order": 3,
  },
]
`;

exports[`tests headliners class return headliners sorted by customOrderExceptHeadliners type 1`] = `
Array [
  Object {
    "artist": "Snoop Dogg",
    "day": 1,
    "decorator": "promoted",
    "firstOnLine": Array [
      "headliners",
      "lineup",
    ],
  },
  Object {
    "artist": "Fatboy Slim",
    "day": 1,
    "decorator": "promoted",
  },
  Object {
    "artist": "Faithless",
    "day": 1,
    "decorator": "promoted",
    "firstOnLine": "lineup",
  },
  Object {
    "artist": "The White Stripes",
    "day": 2,
    "firstOnLine": Array [
      "headliners",
      "lineup",
    ],
  },
  Object {
    "artist": "Lauryn Hill",
    "day": 2,
  },
  Object {
    "artist": "Underworld",
    "day": 2,
    "firstOnLine": "lineup",
  },
]
`;
